src/engine/source/logicexpr/interface/logicexpr/parser.hpp

```diff
@@ -0,0 +1,368 @@
+// https://stackoverflow.com/questions/29634992/shunting-yard-validate-expression
+struct syntaxChecker
+{
+
+    // Could be optimized by taking advantage of the fact that shunting-yard
+    // algorithm is already doing some comparisons, is as it is for modularity
+
+    // Two states:
+    // - 0: expect operand, unary operator or parenthesis
+    // - 1: expect binary operator
+    bool m_state;
+
+    bool expectedOperand() const { return m_state == false; }
+
+    void expectOperand() { m_state = false; }
+
+    bool expectedOperator() const { return m_state == true; }
+
+    void expectOperator() { m_state = true; }
+
+    syntaxChecker()
+        : m_state {false}
+    {
+    }
+
+    // Check if token satisfies the current state, otherwise throw exception
+    void operator()(const Token& token)
+    {
+        // Got term
+        if (token->isTerm())
+        {
+            if (expectedOperator())
+            {
+                throw std::runtime_error(
+                    fmt::format("Unexpected token TERM '{}' at position '{}'", token->text(), token->pos()));
+            }
+
+            expectOperator();
+            return;
+        }
+
+        // Got unary operator
+        if (token->isUnaryOperator())
+        {
+            if (expectedOperator())
+            {
+                throw std::runtime_error(fmt::format("Unexpected unary operator 'NOT' at position '{}'", token->pos()));
+            }
+
+            // Still wanting operand
+            return;
+        }
+
+        // Got binary operator
+        if (token->isBinaryOperator())
+        {
+            if (expectedOperand())
+            {
+                throw std::runtime_error(
+                    fmt::format("Unexpected binary operator '{}' at position '{}'", token->text(), token->pos()));
+            }
+
+            expectOperand();
+            return;
+        }
+
+        // Got parenthesis open
+        if (token->isParenthesisOpen())
+        {
+            if (expectedOperator())
+            {
+                throw std::runtime_error(fmt::format("Unexpected parenthesis '(' at position '{}'", token->pos()));
+            }
+
+            // Still wanting operand
+            return;
+        }
+
+        // Got parenthesis close
+        if (token->isParenthesisClose())
+        {
+            if (expectedOperand())
+            {
+                throw std::runtime_error(fmt::format("Unexpected parenthesis ')' at position '{}'", token->pos()));
+            }
+
+            // Still wanting operator
+            return;
+        }
+    }
+};
+
+/**
+ * @brief Transforms a queue of tokens in infix notation into a stack of tokens in postfix notation using the Shunting-Yard algorithm.
+ *
+ * @param infix Queue of tokens in infix notation.
+ * @return std::stack<Token> Stack of tokens in postfix notation.
+ * @throw std::runtime_error if the parenthesis are not balanced.
+ */
+std::stack<Token> infixToPostfix(std::queue<Token>& infix)
```

Comment:
nitpick - desing: Code duplication could be reduced by using lambda functions to handle operators and closing parentheses.

```
    auto handleOperator = [&](const Token& token) {
        while (!operatorStack.empty() && !operatorStack.top()->isParenthesisOpen() &&
               operatorStack.top()->getPtr<OpToken>()->precedence() >= token->getPtr<OpToken>()->precedence())
        {
            postfix.push(std::move(operatorStack.top()));
            operatorStack.pop();
        }
        operatorStack.push(std::move(token));
    };
    
    auto handleClosingParenthesis = [&]() {
        while (!operatorStack.empty() && !operatorStack.top()->isParenthesisOpen()) {
            postfix.push(std::move(operatorStack.top()));
            operatorStack.pop();
        }
        if (operatorStack.empty()) {
            throw std::runtime_error("Parenthesis are not balanced");
        }
        operatorStack.pop();
    };

```